src/simtools/corsika/corsika_histograms.py

```diff
@@ -630,1240 +144,401 @@ def _fill_histograms(self, photons, rotation_around_z_axis=None, rotation_around
-        bins: float
-            Number of bins for the histogram.
-        hist_range: 2-tuple
-            Tuple to define the range of the histogram.
-
-        Returns
-        -------
-        numpy.ndarray
-            The counts of the histogram.
-        numpy.array
-            Edges of the histogram.
-
-
-        Raises
-        ------
-        KeyError:
-            If key is not valid.
-        """
-        if key not in self.all_event_keys:
-            msg = f"key is not valid. Valid entries are {self.all_event_keys}"
-            self._logger.error(msg)
-            raise KeyError
-        hist, bin_edges = np.histogram(
-            self.event_information[key].value,
-            bins=bins,
-            range=hist_range,
-        )
-        return hist, bin_edges
-
-    def event_2d_histogram(self, key_1, key_2, bins=50, hist_range=None):
-        """
-        Create a 2D histogram for the all events using key_1 and key_2 as parameters.
+        return hist_2d
 
-        Valid keys are stored in self.all_event_keys (CORSIKA defined).
+    def _update_distributions(self):
+        """Update the distributions dictionary with the histogram values and bin edges."""
+        self._normalize_density_histograms()
 
-        Parameters
-        ----------
-        key_1: str
-            The information from which to build the histogram, e.g. total_energy, zenith or
-            first_interaction_height.
-        key_2: str
-            The information from which to build the histogram, e.g. total_energy, zenith or
-            first_interaction_height.
-        bins: float
-            Number of bins for the histogram.
-        hist_range: 2-tuple
-            Tuple to define the range of the histogram.
+        for key, value in self.hist.items():
+            value["input_file_name"] = str(self.input_file)
+            if value["is_1d"]:
+                value["hist_values"], value["x_bin_edges"], value["uncertainties"] = (
+                    self.get_hist_1d_projection(key, value)
+                )
+            else:
+                (
+                    value["hist_values"],
+                    value["x_bin_edges"],
+                    value["y_bin_edges"],
+                    value["uncertainties"],
+                ) = self.get_hist_2d_projection(value["histogram"])
+
+    def _normalize_density_histograms(self):
+        """Normalize the density histograms by the area of each bin."""
+
+        def normalize_histogram(hist, bin_areas):
+            view = hist.view()
+            if hasattr(view, "dtype") and view.dtype.names == ("value", "variance"):
+                view["value"] /= bin_areas
+                view["variance"] /= bin_areas**2
+            else:
+                hist /= bin_areas
+
+        def normalize_histogram_1d(hist, bin_areas):
+            view = hist.view()
+            if hasattr(view, "dtype") and view.dtype.names == ("value", "variance"):
+                for i, area in enumerate(bin_areas):
```

Comment:
consider removing the loop, because view["value"] and bin_areas are NumPy arrays.

Reply:
Fixed.